src/pveclib/vec_int32_ppc.h

```diff
@@ -1419,4 +1300,257 @@ vec_srwi (vui32_t vra, const unsigned int shb)
   return (vui32_t) result;
 }
 
+/** \brief \copybrief vec_int64_ppc.h::vec_vmadd2euw()
+ *
+ * \note this implementation exists in
+ * \ref vec_int64_ppc.h::vec_vmadd2euw()
+ * as it requires vec_addudm().
+ */
+static inline vui64_t
+vec_vmadd2euw (vui32_t a, vui32_t b, vui32_t c, vui32_t d);
+
+/** \brief \copybrief vec_int64_ppc.h::vec_vmadd2ouw()
+ *
+ * \note this implementation exists in
+ * \ref vec_int64_ppc.h::vec_vmadd2ouw()
+ * as it requires vec_addudm().
+ */
+static inline vui64_t
+vec_vmadd2ouw (vui32_t a, vui32_t b, vui32_t c, vui32_t d);
+
+/** \brief \copybrief vec_int64_ppc.h::vec_vmaddeuw()
+ *
+ * \note this implementation exists in
+ * \ref vec_int64_ppc.h::vec_vmaddeuw()
+ * as it requires vec_addudm().
+ */
+static inline vui64_t
+vec_vmaddeuw (vui32_t a, vui32_t b, vui32_t c);
+
+/** \brief \copybrief vec_int64_ppc.h::vec_vmaddouw()
+ *
+ * \note this implementation exists in
+ * \ref vec_int64_ppc.h::vec_vmaddouw()
+ * as it requires vec_addudm().
+ */
+static inline vui64_t
+vec_vmaddouw (vui32_t a, vui32_t b, vui32_t c);
+
+/** \brief \copybrief vec_int64_ppc.h::vec_vmsumuwm()
+ *
+ * \note this implementation exists in
+ * \ref vec_int64_ppc.h::vec_vmsumuwm()
+ * as it requires vec_addudm().
+ */
+static inline vui64_t
+vec_vmsumuwm (vui32_t a, vui32_t b, vui64_t c);
+
+/** \brief Vector Multiply Even Unsigned words.
+ *
+ * Multiple the even words of two vector unsigned int values and return
+ * the unsigned long product of the even words.
+ *
+ * For POWER8 and later we can use the vmuleuw instruction.
+ * But for POWER7 and earlier we have to construct word multiplies
+ * from two halfword multiplies (vmuleuh and vmulouh). Then sum the
+ * partial products for the final doubleword results. This is
+ * complicated by the fact that vector add doubleword is not available
+ * for POWER7. So we need to construct the doubleword add from
+ * Vector Add Unsigned Word Modulo (vadduwm) and
+ * Vector Add and Write Carry-Out Unsigned Word (vaddcuw) with
+ * shift double quadword to reposition the low word carry and
+ * a final vadduwm to complete the carry propagation for the
+ * doubleword add.
+ *
+ * \note This implementation is NOT endian sensitive and the function
+ * is stable across BE/LE implementations.
+ *
+ * |processor|Latency|Throughput|
+ * |--------:|:-----:|:---------|
+ * |power8   |   7   | 2/cycle  |
+ * |power9   |   7   | 2/cycle  |
```

Comment:
No power7 data? (I doubt there's too much interest, but you did go to a lot of trouble to support POWER7.

Reply:
There problem is there is no "public" version of the POWER7 book 4 (AKA users manual). These do exist for the POWER8/9 on OpenPOWER.org.

Before I retired I had access to the IBM confidential Book 4s, but i purged those during my last week at IBM.

I don't want to publish guesses and so choose to not publish any cycle times for P7.